command/markad.1
.\" ** The above line should force tbl to be a preprocessor **
.\" Man page for markad
.\" 
.\" Copyright (C) 2012 Jochen Dolze
.\" 
.\" You may distribute under the terms of the GNU General Public
.\" License as specified in the file COPYING that comes with the
.\" VDR distribution.
.\" 
.\" $Id: markad.1 1.0 2012/05/25 22:33:34 martinkg Exp $
.\" 
.TH "markad" "1" "25 September 2024" "24.2.0" "A program for the Video Disk Recorder"
.SH "NAME"
MarkAd \- marks advertisements in VDR recordings.
.SH "SYNOPSIS"
.B markad
[options]
.SH "DESCRIPTION"
.B MarkAd
marks advertisements in VDR recordings.
.SH "OPTIONS"
.TP 
Usage: markad [options] <cmd> <recording>
.TP 

.BI \-b\ ,\ \-\-background
\fImarkad\fR runs as a background process;
this will be automatically set if called with "after"
.TP 

.BI \-l\ ,\ \-\-logocachedir
directory where predefined logos were stored, default \fI/var/lib/markad\fR
.RS
Note: Channels sometimes change their logo. If logos kept in the cache are out of date, this will result in bad mark results.
.RE
.TP 

.BI \-\-autologo= <option>
<option>   0 = use logo only from logo cache directory
           1 = extract logo from recording, if fails use logo from logo cache
           2 = use logo from logo cache directory, if missing extract logo from recording (default)
.RS
If no suitable logo file is found, markad will try to find the logo in the video and store the result in the recording directory.
.RE
.TP 

.BI \-\-fulldecode
decode all video frame types
use it only on powerful CPUs, it will double overall run time
.TP

.BI \-\-cut
cut video based on marks and store it in the recording directory
there are no splits of the target file, make sure your filesystem can handle big files
.TP


.BI \-\-smartencode
this option is only available on command line usage
requires --cut
re-encode only short before and short after cut position
.TP

.BI \-\-fullencode=<streams>
this option is only available for command line usage
requires --cut
full re-encode video generated by --cut
use it only on powerful CPUs, it will double overall run time
<streams>  all  = keep all video and audio streams of the recording
           best = only encode best video and best audio stream, drop rest
.TP

.BI \-p\ ,\ \-\-priority= <priority>
software priority of markad when running in background
.RS
<priority>  from \-20...19, default 19
.RE
.TP 

.BI \-r\ ,\ \-\-ioprio= <class>[,<level>]
I/O priority of markad when running in background
.RS
<class>     1 = realtime,   <level> from 0..7, default 4
            2 = besteffort, <level> from 0..7, default 4
            3 = idle (default)
.RE
.RS
Make sure your I/O scheduler supports scheduling priorities and classes (e.g. BFQ or CFQ).
.RE
.TP 

.BI \-v\ ,\ \-\-verbose
increments loglevel by one, can be given multiple times
.TP 

.BI \-B\ ,\ \-\-backupmarks
make a backup of existing marks
.TP 

.BI \-L\ ,\ \-\-extractlogo= <direction>[,width[,height]]
this parameter is deprecated, use autologo instead
.RS
extracts logo to recording directory as pgm files (must be renamed)
.RS
<direction>  0 = top left,    1 = top right
             2 = bottom left, 3 = bottom right
             [width]  range from 50 to 480, default 192 (SD) / 288 (HD)
             [height] range from 20 to 250, default 100 (SD) / 180 (HD)
.RE
.RE
.TP 

.BI \-O\ ,\ \-\-OSD
markad sends an OSD message for start and end
.TP 

.BI \-R\ ,\ \-\-log2rec
write log files into recording directory
.TP 

.BI \-T\ ,\ \-\-threads= <number>
number of threads used for decoding, max. 16
(default is the number of CPUs)
.TP 

.BI \-V\ ,\ \-\-version
print version info and exit
.TP 

.BI \-\-loglevel= <level>
sets log level to the specified value
.RS
<level>      1 = error
             2 = info
             3 = debug
             4 = trace
.RE
.TP 

.BI \-\-markfile= <markfilename>
set a different marks filename
.TP 

.BI \-\-nopid
disables creation of \fImarkad.pid\fR file in the recording directory
.TP 

.BI \fB\-\-online[=1|2] 
.RS
only used by markad plugin internal call, do not use it for command line start
.RE
.RS
default is 1
.RE
.RS
start markad immediately when called with "before" as cmd
.RE
.RS
If online is 1, markad starts online for live recordings
only, online=2 starts markad online for every recording.
Live recordings are identified by having a '@' in the
filename, so the entry 'Mark instant recording' in the menu
\fISetup\fR > \fIRecording\fR of the VDR should be set to 'yes'.
.RE
.TP 
.BI \-\-svdrphost= "\fR<ip/hostname> " "\fR(default is 127.0.0.1)"
IP or hostname of a remote VDR for OSD messages
.TP

.BI \-\-ac3reencode
re-encode AC3 stream to fix low audio level of cut video on same devices
requires --cut
use only if really needed, as this consumes significant CPU resources
.RS
.RE
.TP 

.BI \-\-hwaccel= <hardware_acceleration_methode>
use hardware acceleration for decoding
all methods supported by FFmpeg (ffmpeg -hide_banner -hwaccels)
e.g.: vaapi, vdpau, vulkan, cuda ...
.TP

.BI \-\-hwaccel-device= <path_to_device>
path to the hardware acceleration device to use
useful for multi-GPU systems to bypass the default device
e.g.: /dev/dri/renderD129
.TP
.BI \-\-vaapi-driver= <driver_name>
force a specific VA-API driver name via FFmpeg dictionary
bypasses dropped environment variables when running via sudo
e.g.: i965, iHD
.TP

.BI \-\-perftest
run decoder performance test and compare software and hardware decoder
.TP

.BI \-\-vps
use VPS events from markad.vps to optimize start and stop marks
.TP

.BI \-\-svdrpport= \fR<port>\fR  "  ( default is 6419 ) "
 port of a remote VDR for OSD messages
.TP
\fB<cmd>\fR one of:
.RE
.RS
\-            dummy parameter if called directly
.RE
.RS
nice         runs markad directly and with nice(19)
.RE
.RS
after        markad started by VDR after the recording is complete
             only used by markad plugin internal call, do not use it for command line start
.RE
.RS
before       markad started by VDR before the recording is complete, only valid together with --online
             only used by markad plugin internal call, do not use it for command line start
.RE
.RS
edited       markad started by VDR in edit function and exits immediately
             only used by markad plugin internal call, do not use it for command line start
.RE
<record>     is the name of the directory where the recording is stored
.SH "EXIT STATUS"
.TP
.B 0
Successful program execution
.TP
.B 1
An non-recoverable error has been detected, markad has given up. See stderr, syslog or log file in the recording directory for more information

.SH "AUTHOR"
Written by Jochen Dolze <vdr@dolze.de>
.SH "REPORTING BUGS"
Report bugs to https://github.com/kfb77/vdr\-plugin\-markad/issues
.SH "COPYRIGHT"
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
See the file COPYING for more information.


HISTORY
VDR Plugin 'markad' Revision History
----------------------------------

2026-07-04: Version 4.3.0
- new feature: parameter --hwaccel-device (specify GPU Device) and --vaapi-driver (specify VAAPI Driver to use)
- fix logo extraction in case of 'Neue Folge'
  thx to Schwabe@www.vdr-portal.de for reporting
- some minor bug fixes and optimizations, see git

2026-05-13: Version 4.2.22
- fix running state if markad start is disabled
  thx to MarkusE@www.vdr-portal.de for reporting
- some minor bug fixes and optimizations, see git

2026-04-09: Version 4.2.21
- cOverlapAroundAd::Process(): fix overlap detection of H.264 SD video
  Thx to Mike838@www.vdr-portal.de for reporting
- some minor bug fixes and optimizations, see git

2026-04-05: Version 4.2.20
- fix encoder nullptr crash
- fix crash with hborder double episode
- some minor bug fixes and optimizations, see git

2026-03-02: Version 4.2.19
- fix channel name not detected with new glib on archlinux
  thx to androderdon@www.vdr-portal.de for reporting
- rewrite markad state monitoring
  thx to FireFly@www.vdr-portal.de for finding a design flaw

2026-02-16: Version 4.2.18
- Makefile: fix rebuilding of markad-standalone.cpp in git tree despite no changes
  thx to SHofmann@www.vdr-portal.de for contribution
- fix svdrpsend plug mark command not working with setup mark during:
  thx to FireFly@www.vdr-portal.de for reporting
- fix crash of svdrpsend plug mark command with missing filename

2025-12-24: Version 4.2.17
- cMarkAdStandalone::Check_HBORDERSTART(): optimize detection of invalid hborder
  thx to FireFly@www.vdr-portal.de for reporting
- send OSD start before logo search
  thx to FireFly@www.vdr-portal.de for reporting
- fix missing analyse of each frame with --fulldecode of H.264 interlaced
- improvements to the recognition of UHD logos
- some minor bug fixes and optimizations, see git

2025-11-30: Version 4.2.16
- new feature: add config parameter to quiet plugin logging if VDR runs with debug log level
- declare parameter --extractlogo as deprecated
- fix syslog priority
- Makefile: fix timestamp of version.h
  thx to FireFly@www.vdr-portal.de for contribution
- fix crash with parameter --extractlogo
  thx to FireFly@www.vdr-portal.de for contribution
- fix status screen position
  thx to FireFly@www.vdr-portal.de for contribution
- read aspect ratio additionally from the F line of the info file
  thx to FireFly@www.vdr-portal.de for contribution
- markad.conf: fix and update documentation
  thx to FireFly@www.vdr-portal.de for reporting
- cLogoDetect::LoadLogoPlane(): fix use after free
  thx to FireFly@www.vdr-portal.de for reporting
- some minor bug fixes and optimizations, see git

2025-07-31: Version 4.2.15
- init threads with number of CPUs
  workaround for problems with libav CPU count detection on aarch64
  thx to Joe_D@www.vdr-portal.de for contribution
- ioprio_set/ioprio_get: add aarch64
  thx to Joe_D@www.vdr-portal.de for contribution
- plugin setup: add hwaccel drm and opencl
  thx to Joe_D@www.vdr-portal.de for contribution
- cStatusMarkAd::Replaying(): fix debug log message
  thx to Joe_D@www.vdr-portal.de for contribution
- some minor bug fixes and optimizations, see git

2025-06-03: Version 4.2.14
- remove useless error message "unknown recording" if markad start disabled
- some minor bug fixes and optimizations, see git

2025-05-07: Version 4.2.13
- some minor bug fixes and optimizations, see git

2025-04-03: Version 4.2.12
- fix logo size check for some channels now without "HD" in logo
- some minor bug fixes and optimizations, see git

2025-03-29: Version 4.2.11
- some minor bug fixes and optimizations, see git

2025-02-25: Version 4.2.10
- fix missing markad finished OSD message (thx to utiltiy@www.vdr-portal.de for reporting)
- some minor bug fixes and optimizations, see git

2025-02-03: Version 4.2.9
- fix not accurate start position with --cut in rare cases
- some minor bug fixes and optimizations, see git

2024-12-24: Version 4.2.8
- cleanup pid file after abort
- fix crash after abort
- some minor bug fixes and optimizations, see git

2024-11-30: Version 4.2.7
- prevent double check of closing credits
- improve performance of logo extraction from recording
- some minor bug fixes and optimizations, see git

2024-11-07: Version 4.2.6
- error exit on invalid use of --online parameter (thx to nobanzai@www.vdr-portal.de for reporting)
- remove shutdown debug message (requested by shofmann@www.vdr-portal.de and MarkusE@www.vdr-portal.de)
- fix start of markad from svdrpsend if start before/after recording is disabled
- some minor bug fixes and optimizations, see git

2024-10-31: Version 4.2.5
- plugin: remove use of MainThreadHook()
- fix crash from missing codec
- detect black screen despite permanent news ticker
- some minor bug fixes and optimizations, see git

2024-10-18: Version 4.2.4
- fix parameter evaluation (thx to hopsi@www.vdr-portal.de and beinhart@www.vdr-portal.de for reporting)
- remove check of APIVERSION in Makefile, no longer working with new VDR APIVERSION value (thx to hopsi@www.vdr-portal.de for reporting)
- some minor bug fixes and optimizations, see git

2024-10-13: Version 4.2.3
- fix encoder color parameters for all codecs (thx to wmautner@www.vdr-portal.de for reporting)
- some minor bug fixes and optimizations, see git

2024-10-05: Version 4.2.2
- hwaccel: use h264_nvenc encoder after decoding with cuda (thx to wmautner@www.vdr-portal.de for reporting)
- encoder: fix missing color space
  this fixed post processing of cut result with FFmpeg transcoding to H.265 (thx to wmautner@www.vdr-portal.de for reporting and a lot of testing)
- some minor bug fixes and optimizations, see git

2024-09-29: Version 4.2.1
- reformat man page (thx to shofmann@www.vdr-portal.de)
- ignore VPS start event after recording interruption (thx to pmrb@www.vdr-portal.de for reporting)
- special logo treatment for some french channel (thx to pmrb@www.vdr-portal.de for reporting)

plugin/markad.cpp
/*
 * markad.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 */

#include <unistd.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <vdr/plugin.h>

#include "markad.h"
#include "debug.h"

#define DEF_BINDIR "/usr/bin"
#define DEF_LOGODIR "/var/lib/markad"


cPluginMarkAd::cPluginMarkAd(void) {
    // Initialize any member variables here.
    // DON'T DO ANYTHING ELSE THAT MAY HAVE SIDE EFFECTS, REQUIRE GLOBAL
    // VDR OBJECTS TO EXIST OR PRODUCE ANY OUTPUT!
    statusMonitor = nullptr;

    bindir = strdup(DEF_BINDIR);
    ALLOC(strlen(bindir)+1, "bindir");

    logodir = strdup(DEF_LOGODIR);
    ALLOC(strlen(logodir)+1, "logodir");

    title[0]                = 0;
    setup.ProcessDuring     = PROCESS_NEVER;
    setup.whileRecording    = true;
    setup.whileReplaying    = true;
    setup.OSDMessage        = false;
    setup.verbosePlugin     = false;
    setup.verboseMarkad     = false;
    setup.HideMainMenuEntry = false;
    setup.Log2Rec           = false;
    setup.LogoOnly          = true;
    setup.DeferredShutdown  = true;
    setup.hwaccel           = 0;
}


cPluginMarkAd::~cPluginMarkAd() {
    // Clean up after yourself!
    if (statusMonitor) {
        FREE(sizeof(*statusMonitor), "statusMonitor");
        delete statusMonitor;
    }
    if (bindir) {
        FREE(strlen(bindir)+1, "bindir");
        free(bindir);
    }
    if (logodir) {
        FREE(strlen(logodir)+1, "logodir");
        free(logodir);
    }
    if (setup.aStopOffs) {
        FREE(strlen(setup.aStopOffs)+1, "setup.aStopOffs");
        free(setup.aStopOffs);
    }
    if (setup.LogLevel) {
        FREE(strlen(setup.LogLevel)+1, "setup.LogLevel");
        free(setup.LogLevel);
    }
#ifdef DEBUG_MEM
    memList();
    memClear();
#endif
}


const char *cPluginMarkAd::CommandLineHelp(void) {
    // Return a string that describes all known command line options.
    return "  -b DIR,   --bindir=DIR         use DIR as location for markad executable\n"
           "                                 (default: /usr/bin)\n"
           "  -l DIR    --logocachedir=DIR   use DIR as location for markad logos\n"
           "                                 (default: /var/lib/markad)\n"
           "            --loglevel=<level>   sets log level of started markad process (standalone, not the plugin) to the specified value\n"
           "                                 <level>: 1=error 2=info 3=debug 4=trace\n"
           "            --astopoffs=<value>  assumed stop offset (to start + length of broadcast) in seconds, range from 0 to 240\n"
           "                                 (default is 0)\n"
           "            --cut                cut video based on marks and store it in the recording directory)\n"
           "            --ac3reencode        re-encode AC3 stream to fix low audio level of cutted video on same devices\n"
           "                                 requires --cut\n"
           "            --autologo=<option>  0 = disable, only use logos from logo cache directory\n"
           "                                 1 = deprecated, do not use\n"
           "                                 2 = enable (default)\n"
           "                                     if there is no suitable logo in the logo cache directory markad will\n"
           "                                     try to find the logo from recording and store it in the recording directory\n"
           "                                     If this option is set you can not configure this feature from the VDR menu\n";
}


bool cPluginMarkAd::ProcessArgs(int argc, char *argv[]) {
    // Command line argument processing
    static struct option long_options[] = {
        { "bindir",       required_argument, nullptr, 'b'},
        { "logocachedir", required_argument, nullptr, 'l'},
        { "loglevel",     required_argument, nullptr, '1'},
        { "astopoffs",    required_argument, nullptr, '2'},
        { "cut",          no_argument,       nullptr, '3'},
        { "ac3reencode",  no_argument,       nullptr, '4'},
        { "autologo",     required_argument, nullptr, '5'},
        { nullptr, 0, nullptr, 0 }
    };

    int c;
    while ((c = getopt_long(argc, argv, "b:l:", long_options, nullptr)) != -1) {
        switch (c) {
        case 'b':
            if ((access(optarg,R_OK | X_OK)) != -1) {
                if (bindir) {
                    FREE(strlen(bindir)+1, "bindir");
                    free(bindir);
                }
                bindir=strdup(optarg);
                ALLOC(strlen(bindir)+1, "bindir");
            }
            else {
                fprintf(stderr,"markad: can't access bin directory: %s\n", optarg);
                return false;
            }
            break;
        case 'l':
            if ((access(optarg,R_OK)) != -1) {
                if (logodir) {
                    FREE(strlen(logodir)+1, "logodir");
                    free(logodir);
                }
                logodir=strdup(optarg);
                ALLOC(strlen(logodir)+1, "logodir");
            }
            else {
                fprintf(stderr,"markad: can't access logo directory: %s\n", optarg);
                return false;
            }
            break;
        case '1':
            loglevel = atoi(optarg);
            break;
        case '2':
            astopoffs = atoi(optarg);
            break;
        case '3':
            MarkadCut = true;
            break;
        case '4':
            ac3ReEncode = true;
            break;
        case '5':
            autoLogoConf = atoi(optarg);
            break;
        default:
            return false;
        }
    }
    return true;
}


bool cPluginMarkAd::Initialize(void) {
    // Initialize any background activities the plugin shall perform.
    isyslog("markad: start markad plugin with loglevel %d", logLevel);
    char *path;
    if (asprintf(&path, "%s/markad", bindir) == -1) return false;
    ALLOC(strlen(path)+1, "path");

    struct stat statbuf;
    if (stat(path, &statbuf) == -1) {
        esyslog("markad: cannot find %s, please install", path);

        FREE(strlen(path)+1, "path");
        free(path);
        return false;
    }
    FREE(strlen(path)+1, "path");
    free(path);

    DebugLog("cPluginMarkAd::Initialize(): create status monitor");
    statusMonitor = new cStatusMarkAd(bindir, logodir, &setup);
    ALLOC(sizeof(*statusMonitor), "statusMonitor");
    return (statusMonitor!=nullptr);
}


bool cPluginMarkAd::Start(void) {
    // Start any background activities the plugin shall perform.
    DebugLog("cPluginMarkAd::Start(): called");
    lastcheck = 0;
    setup.PluginName = Name();
    if (loglevel) {
        if(! asprintf(&setup.LogLevel, " --loglevel=%i ", loglevel)) esyslog("markad: asprintf out of memory");
        ALLOC(strlen(setup.LogLevel)+1, "setup.LogLevel");
    }

    if (astopoffs >= 0) {
        if(! asprintf(&setup.aStopOffs, " --astopoffs=%i ", astopoffs)) esyslog("markad: asprintf out of memory");
        ALLOC(strlen(setup.aStopOffs)+1, "setup.aStopOffs");
    }
    setup.MarkadCut    = MarkadCut;
    setup.ac3ReEncode  = ac3ReEncode;
    setup.autoLogoConf = autoLogoConf;
    setup.LogoDir      = logodir;

    return true;
}


void cPluginMarkAd::Stop(void) {
    // Stop any background activities the plugin is performing.
    DebugLog("cPluginMarkAd::Stop(): called");
}


void cPluginMarkAd::Housekeeping(void) {
    // Perform any cleanup or other regular tasks.
    // looks like only called if no recording is running, but that's enough to keep track of whether all the marks have been completed
    // DebugLog("cPluginMarkAd::Housekeeping(): called");
    // refresh markad status
    if (setup.ProcessDuring != PROCESS_NEVER) statusMonitor->RefreshStatus();
}


cString cPluginMarkAd::Active(void) {
    // Return a message string if shutdown should be postponed
    if (statusMonitor->MarkAdRunning() && (setup.DeferredShutdown)) {
        DebugLog("markad still running, shutdown request rejected");
        return tr("markad still running");
    }
    return nullptr;
}


time_t cPluginMarkAd::WakeupTime(void) {
    // Return custom wakeup time for shutdown script
    return 0;
}


cOsdObject *cPluginMarkAd::MainMenuAction(void) {
    // Perform the action when selected from the main VDR menu.
    return new cMenuMarkAd(statusMonitor);  // this should be freed from VDR after meue closed
}


cMenuSetupPage *cPluginMarkAd::SetupMenu(void) {
    // Return the setup menu
    return new cSetupMarkAd(&setup);   // this should be freed from VDR after meue closed
}


bool cPluginMarkAd::SetupParse(const char *Name, const char *Value) {
    // Parse setup parameters and store their values.
    if (!strcasecmp(Name,"Execution"))               setup.ProcessDuring     = atoi(Value);
    else if (!strcasecmp(Name,"useVPS"))             setup.useVPS            = atoi(Value);
    else if (!strcasecmp(Name,"logVPS"))             setup.logVPS            = atoi(Value);
    else if (!strcasecmp(Name,"whileRecording"))     setup.whileRecording    = atoi(Value);
    else if (!strcasecmp(Name,"whileReplaying"))     setup.whileReplaying    = atoi(Value);
    else if (!strcasecmp(Name,"OSDMessage"))         setup.OSDMessage        = atoi(Value);
    else if (!strcasecmp(Name,"svdrPort"))           setup.svdrPort          = atoi(Value);
    else if (!strcasecmp(Name,"VerbosePlugin"))      setup.verbosePlugin     = atoi(Value);
    else if (!strcasecmp(Name,"Verbose"))            setup.verboseMarkad     = atoi(Value);
    else if (!strcasecmp(Name,"HideMainMenuEntry"))  setup.HideMainMenuEntry = atoi(Value)?true:false;
    else if (!strcasecmp(Name,"Log2Rec"))            setup.Log2Rec           = atoi(Value);
    else if (!strcasecmp(Name,"LogoOnly"))           setup.LogoOnly          = atoi(Value);
    else if (!strcasecmp(Name,"DeferredShutdown"))   setup.DeferredShutdown  = atoi(Value);
    else if (!strcasecmp(Name,"AutoLogoExtraction")) setup.autoLogoMenu      = atoi(Value);
    else if (!strcasecmp(Name,"FullDecode"))         setup.fulldecode        = atoi(Value);
    else if (!strcasecmp(Name,"hwaccel"))            setup.hwaccel           = atoi(Value);
    else return false;

    if (setup.verbosePlugin) logLevel = 3;
    else                     logLevel = 2;
    return true;
}


const char *cPluginMarkAd::MainMenuEntry(void) {
    if (setup.HideMainMenuEntry)
        return nullptr;
    else
        return tr("markad status");
}


bool cPluginMarkAd::Service(const char *UNUSED(Id), void *UNUSED(Data)) {
    // Handle custom service requests from other plugins
    return false;
}


const char **cPluginMarkAd::SVDRPHelpPages(void) {
    // Return help text for SVDRP
    static const char *HelpPage[] = {
        "MARK <filename>\n"
        "     Start markad for the recording with the given directory name.",
        "STATUS\n"
        "     show active recordings with running markad",
#ifdef DEBUG_MEM
        "DEBUG_MEM\n"
        "     show current allocated heap memory",
#endif
        nullptr
    };
    return HelpPage;
}


bool cPluginMarkAd::ReadTitle(const char *Directory) {
    usleep(1000000); // wait 1 second
    memset(&title, 0, sizeof(title));
    char *buf;
    if (asprintf(&buf, "%s/info",Directory) == -1) return false;
    ALLOC(strlen(buf)+1, "buf");

    FILE *f;
    f = fopen(buf,"r");
    FREE(strlen(buf)+1, "buf");
    free(buf);
    buf = nullptr;
    if (!f) {
        if (asprintf(&buf, "%s/info.vdr",Directory) == -1) return false;
        ALLOC(strlen(buf)+1, "buf");
        f = fopen(buf,"r");
        FREE(strlen(buf)+1, "buf");
        free(buf);
        if (!f) return false;
    }

    char *line = nullptr;
    size_t length;
    while (getline(&line, &length, f) != -1) {
        ALLOC(strlen(line) + 1, "line");
        if (line[0] == 'T') {
            int result = sscanf(line, "%*c %79c", title);
            if ((result == 0) || (result == EOF)) {
                title[0] = 0;
            }
            else {
                char *lf = strchr(title, 10);
                if (lf) *lf = 0;
                char *cr = strchr(title, 13);
                if (cr) *cr = 0;
            }
        }
    }
    if (line) {
        FREE(strlen(line) + 1, "line");
        free(line);
    }

    fclose(f);
    return (title[0] != 0);
}


cString cPluginMarkAd::SVDRPCommand(const char *Command, const char *Option, int &ReplyCode) {
// Process SVDRP command
    // start markad via SVDRP command, timerVPS will be detected by markad, we don't know this here
    DebugLog("cPluginMarkAd::SVDRPCommand(): command: %s %s", Command, Option ? Option : "(null)");
    if (strcasecmp(Command, "MARK") == 0) {
        if (Option && *Option != '\0') {
            const char *Title = nullptr;
            if (ReadTitle(Option)) Title = reinterpret_cast<char *>(&title);
            sRecording recording;
            if (statusMonitor->Start(Title, Option, &recording)) {
                return cString::sprintf("Started markad for %s", Option);
            }
            else {
                ReplyCode = 451;
                return cString::sprintf("Failed to start markad for %s", Option);
            }
        }
        else {
            ReplyCode = 501;
            return cString::sprintf("Missing filename");
        }
    }
    if (strcasecmp(Command, "STATUS") == 0) {
        statusMonitor->RefreshStatus();
        return statusMonitor->GetStatus();
    }
#ifdef DEBUG_MEM
    if (strcasecmp(Command,"DEBUG_MEM") == 0) {
        return memListSVDR();
    }
#endif
    return nullptr;
}
VDRPLUGINCREATOR(cPluginMarkAd) // Don't touch this!
